jax/interpreters/sharded_jit.py

```diff
@@ -13,262 +13,172 @@
-  device_assignment = onp.array([[d.id for d in devices]])
-  device_assignment = onp.reshape(device_assignment, (-1, num_partitions))
+  assert len(devices) == num_partitions
+  device_assignment = np.array([[d.id for d in devices]])
+  device_assignment = np.reshape(device_assignment, (-1, num_partitions))
   # device_assignment = None  # TODO(skye): replace with default device assignment?
 
-  compiled = built.Compile(
-      compile_options=xb.get_compile_options(nrep, num_partitions, device_assignment),
-      backend=xb.get_backend(None))
+  compiled = xb.get_backend().compile(
+      built, compile_options=xb.get_compile_options(
+          nrep, num_partitions, device_assignment))
 
-  # logging.error("===== hlo:\n%s" % built.as_hlo_text())
+  input_specs = [
+      _partitioned_sharding_spec(num_partitions, parts, aval)
+      for parts, aval in zip(in_parts, abstract_args)]
+  input_indices = [pxla.spec_to_indices(aval.shape, spec)
+                   if spec is not None else None
+                   for aval, spec in zip(abstract_args, input_specs)]
 
-  handle_args = partial(_spatial_partitioned_args, compiled.local_devices(),
-                        device_assignment, partitions[0])
-  handle_outs = _pvals_to_results_handler(nrep, num_partitions, partitions[1],
+  handle_args = partial(pxla.shard_args, compiled.local_devices(),
+                        input_indices)
+  handle_outs = _pvals_to_results_handler(nrep, num_partitions, out_parts,
                                           out_pvals)
   return partial(_execute_spatially_partitioned, compiled, handle_args,
                  handle_outs)
 
 
-def _sharded_jit_translation_rule(c, axis_env, freevar_nodes,
-                                  in_nodes, name_stack, partitions, backend,
-                                  name, call_jaxpr):
-  subc = xb.make_computation_builder("jaxpr_subcomputation")  # TODO(mattjj): name
-  freevars = [subc.ParameterWithShape(c.get_shape(n)) for n in freevar_nodes]
+def _partitioned_sharding_spec(num_partitions: int,
+                               partitions: Optional[Sequence[int]], aval):
+  if aval is core.abstract_unit:
+    return None
 
-  args = []
-  for p, a in zip(partitions[0], in_nodes):
-    subc._builder.set_sharding(_sharding_to_proto(p))
-    args.append(subc.ParameterWithShape(c.get_shape(a)))
-    subc._builder.clear_sharding()
-  # args = [subc.ParameterWithShape(c.get_shape(n)) for n in in_nodes]
-
-  out_nodes = xla.jaxpr_subcomp(subc, call_jaxpr, backend, axis_env, (), freevars, name_stack, *args)
-
-  subc._builder.set_sharding(_sharding_to_proto(partitions[1]))
-  out_tuple = subc.Tuple(*out_nodes)
-  subc._builder.clear_sharding()
+  if partitions is None:
+    return pxla.ShardingSpec(
+        # int(1) because pytype is confused by 1 (???)
+        shards_per_axis=(int(1),) * len(aval.shape),
+        is_axis_materialized=(True,) * len(aval.shape),
+        replication_factor=num_partitions)
+  else:
```

Comment:
Does the XLA partitioner support partially-sharded, partially-replicated arguments? I'm guessing no?

Reply:
You mean like e.g., cut in half then replicate twice? Pretty sure no. At least I'm not sure how you would express that with an OpSharding. I think regular replication (i.e. with pmap) + sharding achieves the same thing though?